pkg/dns/ibm/dns.go

```diff
@@ -0,0 +1,203 @@
+		ApiKey: config.APIKey,
+	}
+	provider := &Provider{}
+
+	for _, zone := range config.Zones {
+		options := &dnsrecordsv1.DnsRecordsV1Options{
+			Authenticator:  authenticator,
+			URL:            "https://api.private.cis.cloud.ibm.com/v1",
+			Crn:            &config.CISCRN,
+			ZoneIdentifier: &zone,
+		}
+
+		dnsService, err := dnsrecordsv1.NewDnsRecordsV1(options)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create a new DNS Service instance: %w", err)
+		}
+		dnsService.EnableRetries(3, 5*time.Second)
+		dnsService.Service.SetUserAgent(config.UserAgent)
+
+		provider.dnsServices[zone] = dnsService
+	}
+
+	if err := validateDNSServices(provider); err != nil {
+		return nil, fmt.Errorf("failed to validate ibm dns services: %w", err)
+	}
+	return provider, nil
+}
+
+// validateDNSServices validates that provider clients can communicate with
+// associated API endpoints by having each client make a get DNS records call.
+func validateDNSServices(provider *Provider) error {
+	var errs []error
+	maxItems := int64(1)
+	for _, dnsService := range provider.dnsServices {
+		opt := dnsService.NewListAllDnsRecordsOptions()
+		opt.PerPage = &maxItems
+		if _, _, err := dnsService.ListAllDnsRecords(opt); err != nil {
+			errs = append(errs, fmt.Errorf("failed to get dns records: %w", err))
+		}
+	}
+	return kerrors.NewAggregate(errs)
+}
+
+func (p *Provider) Ensure(record *iov1.DNSRecord, zone configv1.DNSZone) error {
+	return p.createOrUpdateDNSRecord(record, zone)
+}
+
+func (p *Provider) Replace(record *iov1.DNSRecord, zone configv1.DNSZone) error {
+	return p.createOrUpdateDNSRecord(record, zone)
+}
+
+func (p *Provider) Delete(record *iov1.DNSRecord, zone configv1.DNSZone) error {
+	if err := validateInputDNSData(record, zone); err != nil {
+		return fmt.Errorf("delete: invalid dns input data: %w", err)
+	}
+	dnsService, ok := p.dnsServices[zone.ID]
+	if !ok {
+		return fmt.Errorf("delete: unknown zone: %v", zone.ID)
+	}
+	opt := dnsService.NewListAllDnsRecordsOptions()
+	opt.SetType(string(record.Spec.RecordType))
+	opt.SetName(record.Spec.DNSName)
+	for _, target := range record.Spec.Targets {
+		opt.SetContent(target)
+		result, response, err := dnsService.ListAllDnsRecords(opt)
+		if err != nil {
+			if response.StatusCode != http.StatusNotFound {
+				return fmt.Errorf("delete: failed to list the dns record: %w", err)
+			}
+			continue
+		}
+		if result == nil || result.Result == nil {
+			return fmt.Errorf("delete: invalid result")
+		}
+		for _, resultData := range result.Result {
+			if resultData.ID == nil {
+				return fmt.Errorf("delete: record id is nil")
+			}
+			delOpt := dnsService.NewDeleteDnsRecordOptions(*resultData.ID)
+			_, delResponse, err := dnsService.DeleteDnsRecord(delOpt)
+			if err != nil && delResponse.StatusCode != http.StatusNotFound {
+				return fmt.Errorf("delete: failed to delete the dns record: %w", err)
+			}
+			if delResponse.StatusCode != http.StatusNotFound {
```

Comment:
Do we need to check to make sure `delResponse` isn't nil?  Come to think of it, we might need to check for nil values where we use the result or response value from `dnsService.ListAllDnsRecords` as well.